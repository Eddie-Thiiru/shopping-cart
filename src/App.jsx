import { useState, createContext } from "react";
import Header from "./components/Header";
import MainSection from "./components/MainSection";
import Footer from "./components/Footer";
import { DataProvider } from "./components/DataProvider";
import "./styles/App.css";

export const CartContext = createContext();

const App = () => {
  const [cartItems, setCartItems] = useState({
    count: 0,
    items: [],
  });

  const isDuplicate = (id) => {
    const array = cartItems.items;

    for (let i = 0; i < array.length; i++) {
      const element = array[i];

      if (element.uniqueId === id) {
        return true;
      }
    }
    return false;
  };

  const addToCart = (data) => {
    const duplicate = isDuplicate(data.uniqueId);

    if (duplicate === false) {
      setCartItems({
        count: cartItems.count + 1,
        items: cartItems.items.concat(data),
      });
    } else {
      return;
    }
  };

  const removeFromCart = (id) => {
    setCartItems({
      count: cartItems.count - 1,
      items: cartItems.items.filter((item) => item.uniqueId !== id),
    });
  };

  const cartData = cartItems.items;

  return (
    <div id="App">
      <Header cartItemsCount={cartItems.count} />
      <DataProvider>
        <CartContext.Provider value={{ cartData, addToCart, removeFromCart }}>
          <MainSection />
        </CartContext.Provider>
      </DataProvider>
      <Footer />
    </div>
  );
};

export default App;
