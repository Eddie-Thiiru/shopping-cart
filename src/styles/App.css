#App {
  height: 100%;
  width: 100%;
  display: grid;
  grid-template-rows: 100px minmax(800px, 1fr) 400px;
}

.header {
  position: sticky;
  width: 100%;
  top: 0;
  /* background-color: rgba(0, 0, 0, 0.8); */
  background: linear-gradient(to bottom, rgb(0, 0, 0), rgba(0, 0, 0, 0.8));
  padding: 10px;
  display: grid;
  grid-template-columns: 1fr 2fr;
  align-items: center;
  gap: 20px;
  z-index: 1;
}

.mainSection {
  padding: 10px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.footer {
  width: 100%;
  padding: 10px;
  display: grid;
  grid-template-rows: repeat(3, 1fr);
  justify-content: center;
  align-items: center;
}

.nav {
  padding: 10px;
  display: flex;
  gap: 10px;
}

.nav > div {
  margin-left: auto;
}

.footer > div {
  display: flex;
  flex-direction: column;
  justify-items: center;
  align-items: center;
  gap: 20px;
}

.footerAbout,
.footerSocial,
.footerPolicies {
  display: flex;
  gap: 30px;
}

.footerCopyright {
  display: flex;
  flex-direction: column;
  gap: 5px;
}

.apiHyperlink img {
  height: 50px;
}

.apiHyperlink a {
  display: flex;
  align-items: center;
  gap: 10px;
}

ul {
  list-style: none;
}

h1 {
  font-size: 2.5em;
  line-height: 1.1;
}

img {
  object-fit: cover;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}
