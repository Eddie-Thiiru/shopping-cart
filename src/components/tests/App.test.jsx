import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import App from "../../App";

describe("App Component", () => {
  it("should render app", () => {
    const { container } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>,
    );
    screen.debug();

    expect(container).toMatchSnapshot();
  });

  // it("renders home content after navigation click", async () => {
  //   const user = userEvent.setup();

  //   render(
  //     <MemoryRouter initialEntries={[{ pathname: "/" }]}>
  //       <App />
  //     </MemoryRouter>,
  //   );

  //   const link = screen.getByRole("link", { name: "Home" });

  //   await user.click(link);

  //   screen.debug();
  // });
});
