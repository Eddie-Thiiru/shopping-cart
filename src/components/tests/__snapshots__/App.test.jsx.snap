// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > should render app 1`] = `
<div>
  <div
    id="App"
  >
    <div
      class="header"
    >
      <h1>
        UltimateGames
      </h1>
      <div
        class="nav"
      >
        <a
          href="/homepage"
        >
          Home
        </a>
        <a
          href="/games/collection"
        >
          Browse
        </a>
        <div>
          <a
            href="/cart"
          >
            Cart
          </a>
          <span>
            0
          </span>
        </div>
      </div>
    </div>
    <div
      class="mainSection"
    />
    <div
      class="footer"
    >
      <p>
        test test test test test
      </p>
    </div>
  </div>
</div>
`;
